File: client-api/src/test/java/org/ostelco/topup/api/resources/ProfileResourceTest.java
package org.ostelco.topup.api.resources;

import io.dropwizard.auth.AuthValueFactoryProvider;
import io.dropwizard.testing.junit.ResourceTestRule;
import io.vavr.control.Either;
import io.vavr.control.Option;
import org.glassfish.jersey.test.grizzly.GrizzlyWebTestContainerFactory;
import org.junit.ClassRule;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.ostelco.prime.client.api.model.Profile;
import org.ostelco.topup.api.auth.AccessTokenPrincipal;
import org.ostelco.topup.api.core.Error;
import org.ostelco.topup.api.db.SubscriberDAO;
import org.ostelco.topup.api.util.AccessToken;
import org.ostelco.topup.api.util.AuthDynamicFeatureFactory;

import javax.ws.rs.client.Entity;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Profile API tests.
 *
 */
public class ProfileResourceTest {

    private static final SubscriberDAO DAO = mock(SubscriberDAO.class);

    private final String email = "boaty@internet.org";
    private static final String key = "secret";
    private final String name = "Boaty McBoatface";
    private final String address = "Storvej 10";
    private final String postCode = "132 23";
    private final String city = "Oslo";

    private final Profile profile = new Profile(email);

    @ClassRule
    public static final ResourceTestRule RULE = ResourceTestRule.builder()
        .addResource(AuthDynamicFeatureFactory.createInstance(key))
        .addResource(new AuthValueFactoryProvider.Binder<>(AccessTokenPrincipal.class))
        .addResource(new ProfileResource(DAO))
        .setTestContainerFactory(new GrizzlyWebTestContainerFactory())
        .build();

    @Test
    public void getProfile() throws Exception {
        ArgumentCaptor<String> arg = ArgumentCaptor.forClass(String.class);

        when(DAO.getProfile(arg.capture())).thenReturn(Either.right(profile));

        Response resp = RULE.target("/profile")
            .request()
            .header("Authorization", String.format("Bearer %s", AccessToken.withEmail(email)))
            .get(Response.class);

        assertThat(resp.getStatus()).isEqualTo(Response.Status.OK.getStatusCode());
        assertThat(resp.getMediaType().toString()).isEqualTo(MediaType.APPLICATION_JSON);

        /* Requires that:
               lombok.anyConstructor.addConstructorProperties=true
           is added to the lombok config file ('lombok.config').
           Ref.: lombok changelog for ver. 1.16.20. */
        assertThat(resp.readEntity(Profile.class)).isEqualTo(profile);
        assertThat(arg.getValue()).isEqualTo(email);
    }

    @Test
    public void createProfile() throws Exception {
        ArgumentCaptor<String> arg1 = ArgumentCaptor.forClass(String.class);
        ArgumentCaptor<Profile> arg2 = ArgumentCaptor.forClass(Profile.class);

        when(DAO.createProfile(arg1.capture(), arg2.capture()))
            .thenReturn(Option.none());

        Response resp = RULE.target("/profile")
            .request(MediaType.APPLICATION_JSON)
            .header("Authorization", String.format("Bearer %s", AccessToken.withEmail(email)))
            .post(Entity.json("{\n" +
                              "    \"name\": \"" + name + "\",\n" +
                              "    \"address\": \"" + address + "\",\n" +
                              "    \"postCode\": \"" + postCode + "\",\n" +
                              "    \"city\": \"" + city + "\",\n" +
                              "    \"email\": \"" + email + "\"\n" +
                              "}\n"));

        assertThat(resp.getStatus()).isEqualTo(Response.Status.CREATED.getStatusCode());
        assertThat(resp.getMediaType()).isNull();
        assertThat(arg1.getValue()).isEqualTo(email);
        assertThat((arg2.getValue()).getEmail()).isEqualTo(email);
        assertThat((arg2.getValue()).getName()).isEqualTo(name);
        assertThat((arg2.getValue()).getAddress()).isEqualTo(address);
        assertThat((arg2.getValue()).getPostCode()).isEqualTo(postCode);
        assertThat((arg2.getValue()).getCity()).isEqualTo(city);
    }

    @Test
    public void updateProfile() throws Exception {
        ArgumentCaptor<String> arg1 = ArgumentCaptor.forClass(String.class);
        ArgumentCaptor<Profile> arg2 = ArgumentCaptor.forClass(Profile.class);

        String newAddress = "Storvej 10";
        String newPostCode = "132 23";

        when(DAO.updateProfile(arg1.capture(), arg2.capture()))
            .thenReturn(Option.none());

        Response resp = RULE.target("/profile")
            .request(MediaType.APPLICATION_JSON)
            .header("Authorization", String.format("Bearer %s", AccessToken.withEmail(email)))
            .put(Entity.json("{\n" +
                             "    \"name\": \"" + name + "\",\n" +
                             "    \"address\": \"" + newAddress + "\",\n" +
                             "    \"postCode\": \"" + newPostCode + "\",\n" +
                             "    \"city\": \"" + city + "\",\n" +
                             "    \"email\": \"" + email + "\"\n" +
                             "}\n"));

        assertThat(resp.getStatus()).isEqualTo(Response.Status.OK.getStatusCode());
        assertThat(resp.getMediaType()).isNull();
        assertThat(arg1.getValue()).isEqualTo(email);
        assertThat((arg2.getValue()).getEmail()).isEqualTo(email);
        assertThat((arg2.getValue()).getName()).isEqualTo(name);
        assertThat((arg2.getValue()).getAddress()).isEqualTo(newAddress);
        assertThat((arg2.getValue()).getPostCode()).isEqualTo(newPostCode);
        assertThat((arg2.getValue()).getCity()).isEqualTo(city);
    }

    @Test
    public void updateWithIncompleteProfile() throws Exception {
        ArgumentCaptor<String> arg1 = ArgumentCaptor.forClass(String.class);
        ArgumentCaptor<Profile> arg2 = ArgumentCaptor.forClass(Profile.class);

        when(DAO.updateProfile(arg1.capture(), arg2.capture()))
            .thenReturn(Option.of(new Error("No profile found")));

        Response resp = RULE.target("/profile")
            .request(MediaType.APPLICATION_JSON)
            .header("Authorization", String.format("Bearer %s", AccessToken.withEmail(email)))
            .put(Entity.json("{\n" +
                             "    \"name\": \"" + name + "\"\n" +
                             "}\n"));

        assertThat(resp.getStatus()).isEqualTo(Response.Status.NOT_FOUND.getStatusCode());
        assertThat(arg1.getValue()).isEqualTo(email);
    }
}

File: client-api/src/test/kotlin/org/ostelco/prime/client/api/resources/ProfileResourceTest.kt
package org.ostelco.prime.client.api.resources

import arrow.core.Either
import com.fasterxml.jackson.module.kotlin.KotlinModule
import com.nhaarman.mockito_kotlin.argumentCaptor
import io.dropwizard.auth.AuthDynamicFeature
import io.dropwizard.auth.AuthValueFactoryProvider
import io.dropwizard.auth.oauth.OAuthCredentialAuthFilter
import io.dropwizard.jackson.Jackson
import io.dropwizard.testing.junit.ResourceTestRule
import org.assertj.core.api.Assertions.assertThat
import org.glassfish.jersey.test.grizzly.GrizzlyWebTestContainerFactory
import org.junit.Before
import org.junit.ClassRule
import org.junit.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.ostelco.prime.client.api.auth.AccessTokenPrincipal
import org.ostelco.prime.client.api.auth.OAuthAuthenticator
import org.ostelco.prime.client.api.store.SubscriberDAO
import org.ostelco.prime.client.api.util.AccessToken
import org.ostelco.prime.model.Subscriber
import java.util.*
import javax.ws.rs.client.Entity
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

/**
 * Profile API tests.
 *
 */
class ProfileResourceTest {

    private val email = "boaty@internet.org"
    private val name = "Boaty McBoatface"
    private val address = "Storvej 10"
    private val postCode = "132 23"
    private val city = "Oslo"

    private val profile = Subscriber(email)

    @Before
    @Throws(Exception::class)
    fun setUp() {
        `when`(AUTHENTICATOR.authenticate(ArgumentMatchers.anyString()))
                .thenReturn(Optional.of(AccessTokenPrincipal(email)))
    }

    @Test
    @Throws(Exception::class)
    fun getProfile() {
        val arg = argumentCaptor<String>()

        `when`(DAO.getProfile(arg.capture())).thenReturn(Either.right(profile))

        val resp = RULE.target("/profile")
                .request()
                .header("Authorization", "Bearer ${AccessToken.withEmail(email)}")
                .get(Response::class.java)

        assertThat(resp.status).isEqualTo(Response.Status.OK.statusCode)
        assertThat(resp.mediaType.toString()).isEqualTo(MediaType.APPLICATION_JSON)

        assertThat(resp.readEntity(Subscriber::class.java)).isEqualTo(profile)
        assertThat(arg.firstValue).isEqualTo(email)
    }

    @Test
    @Throws(Exception::class)
    fun createProfile() {
        val arg1 = argumentCaptor<String>()
        val arg2 = argumentCaptor<Subscriber>()
        val arg3 = argumentCaptor<String>()


        `when`(DAO.createProfile(arg1.capture(), arg2.capture(), arg3.capture()))
                .thenReturn(Either.right(profile))

        val resp = RULE.target("/profile")
                .request(MediaType.APPLICATION_JSON)
                .header("Authorization", "Bearer ${AccessToken.withEmail(email)}")
                .post(Entity.json("{\n" +
                        "    \"name\": \"" + name + "\",\n" +
                        "    \"address\": \"" + address + "\",\n" +
                        "    \"postCode\": \"" + postCode + "\",\n" +
                        "    \"city\": \"" + city + "\",\n" +
                        "    \"email\": \"" + email + "\"\n" +
                        "}\n"))

        assertThat(resp.status).isEqualTo(Response.Status.CREATED.statusCode)
        assertThat(resp.mediaType.toString()).isEqualTo(MediaType.APPLICATION_JSON)
        assertThat(arg1.firstValue).isEqualTo(email)
        assertThat(arg2.firstValue.email).isEqualTo(email)
        assertThat(arg2.firstValue.name).isEqualTo(name)
        assertThat(arg2.firstValue.address).isEqualTo(address)
        assertThat(arg2.firstValue.postCode).isEqualTo(postCode)
        assertThat(arg2.firstValue.city).isEqualTo(city)
        assertThat(arg3.firstValue).isNull()
    }

    @Test
    @Throws(Exception::class)
    fun createProfileWithReferral() {
        val arg1 = argumentCaptor<String>()
        val arg2 = argumentCaptor<Subscriber>()
        val arg3 = argumentCaptor<String>()

        val referredBy = "foo@bar.com"

        `when`(DAO.createProfile(arg1.capture(), arg2.capture(), arg3.capture()))
                .thenReturn(Either.right(profile))

        val resp = RULE.target("/profile")
                .queryParam("referred_by", referredBy)
                .request(MediaType.APPLICATION_JSON)
                .header("Authorization", "Bearer ${AccessToken.withEmail(email)}")
                .post(Entity.json("{\n" +
                        "    \"name\": \"" + name + "\",\n" +
                        "    \"address\": \"" + address + "\",\n" +
                        "    \"postCode\": \"" + postCode + "\",\n" +
                        "    \"city\": \"" + city + "\",\n" +
                        "    \"email\": \"" + email + "\"\n" +
                        "}\n"))

        assertThat(resp.status).isEqualTo(Response.Status.CREATED.statusCode)
        assertThat(resp.mediaType.toString()).isEqualTo(MediaType.APPLICATION_JSON)
        assertThat(arg1.firstValue).isEqualTo(email)
        assertThat(arg2.firstValue.email).isEqualTo(email)
        assertThat(arg2.firstValue.name).isEqualTo(name)
        assertThat(arg2.firstValue.address).isEqualTo(address)
        assertThat(arg2.firstValue.postCode).isEqualTo(postCode)
        assertThat(arg2.firstValue.city).isEqualTo(city)
        assertThat(arg3.firstValue).isEqualTo(referredBy)
    }

    @Test
    @Throws(Exception::class)
    fun updateProfile() {
        val arg1 = argumentCaptor<String>()
        val arg2 = argumentCaptor<Subscriber>()

        val newAddress = "Storvej 10"
        val newPostCode = "132 23"

        `when`(DAO.updateProfile(arg1.capture(), arg2.capture()))
                .thenReturn(Either.right(profile))

        val resp = RULE.target("/profile")
                .request(MediaType.APPLICATION_JSON)
                .header("Authorization", "Bearer ${AccessToken.withEmail(email)}")
                .put(Entity.json("{\n" +
                        "    \"name\": \"" + name + "\",\n" +
                        "    \"address\": \"" + newAddress + "\",\n" +
                        "    \"postCode\": \"" + newPostCode + "\",\n" +
                        "    \"city\": \"" + city + "\",\n" +
                        "    \"email\": \"" + email + "\"\n" +
                        "}\n"))

        assertThat(resp.status).isEqualTo(Response.Status.OK.statusCode)
        assertThat(resp.mediaType.toString()).isEqualTo(MediaType.APPLICATION_JSON)
        assertThat(arg1.firstValue).isEqualTo(email)
        assertThat(arg2.firstValue.email).isEqualTo(email)
        assertThat(arg2.firstValue.name).isEqualTo(name)
        assertThat(arg2.firstValue.address).isEqualTo(newAddress)
        assertThat(arg2.firstValue.postCode).isEqualTo(newPostCode)
        assertThat(arg2.firstValue.city).isEqualTo(city)
    }

    @Test
    @Throws(Exception::class)
    fun updateWithIncompleteProfile() {
        val resp = RULE.target("/profile")
                .request(MediaType.APPLICATION_JSON)
                .header("Authorization", "Bearer ${AccessToken.withEmail(email)}")
                .put(Entity.json("{\n" +
                        "    \"name\": \"" + name + "\"\n" +
                        "}\n"))

        assertThat(resp.status).isEqualTo(Response.Status.BAD_REQUEST.statusCode)
    }

    companion object {

        val DAO = mock(SubscriberDAO::class.java)
        val AUTHENTICATOR = mock(OAuthAuthenticator::class.java)

        @JvmField
        @ClassRule
        val RULE = ResourceTestRule.builder()
                .setMapper(Jackson.newObjectMapper().registerModule(KotlinModule()))
                .addResource(AuthDynamicFeature(
                        OAuthCredentialAuthFilter.Builder<AccessTokenPrincipal>()
                                .setAuthenticator(AUTHENTICATOR)
                                .setPrefix("Bearer")
                                .buildAuthFilter()))
                .addResource(AuthValueFactoryProvider.Binder(AccessTokenPrincipal::class.java))
                .addResource(ProfileResource(DAO))
                .setTestContainerFactory(GrizzlyWebTestContainerFactory())
                .build()
    }
}
